HW2/P5/physics.pyx

```diff
@@ -113,18 +128,27 @@ cpdef update(FLOAT[:, ::1] XY,
         #
         # SUBPROBLEM 1: parallelize this loop over 4 threads, with static
         # scheduling.
-        for i in range(count):
-            sub_update(XY, V, R, i, count, Grid, grid_spacing)
+        for i in prange(count, chunksize = count/4, schedule='static', num_threads=4):
+            sub_update(XY, V, R, i, count, Grid, grid_spacing, locks)
 
         # update positions
         #
         # SUBPROBLEM 1: parallelize this loop over 4 threads (with static
         #    scheduling).
         # SUBPROBLEM 2: update the grid values.
-        for i in range(count):
+        for i in range(Grid.shape[0]):
+            for j in range(Grid.shape[1]):
+                Grid[i, j] = -1
+        for i in prange(count, chunksize = count/4, schedule='static', num_threads=4):
             for dim in range(2):
                 XY[i, dim] += V[i, dim] * t
 
+        # update grids
+        x_loc = <int>(XY[i, 0]/grid_spacing)
+        y_loc = <int>(XY[i, 1]/grid_spacing)
+        if x_loc<Grid.shape[0] and y_loc<Grid.shape[1]:
```

Comment:
I made separate functions to make sure i was in-bounds since it was used often, but this seems fine too.